lib/src/device_authorization_grant.dart

```diff
@@ -0,0 +1,401 @@
+  /// is not recommended by the OAuth 2.0 spec, and should only be used if the
+  /// server doesn't support Basic authentication.
+  ///
+  /// [RFC 2617]: https://tools.ietf.org/html/rfc2617
+  ///
+  /// [httpClient] is used for all HTTP requests made by this grant, as well as
+  /// those of the [Client] is constructs.
+  ///
+  /// [onCredentialsRefreshed] will be called by the constructed [Client]
+  /// whenever the credentials are refreshed.
+  ///
+  /// The scope strings will be separated by the provided [delimiter]. This
+  /// defaults to `" "`, the OAuth2 standard, but some APIs (such as Facebook's)
+  /// use non-standard delimiters.
+  ///
+  /// By default, this follows the OAuth2 spec and requires the server's
+  /// responses to be in JSON format. However, some servers return non-standard
+  /// response formats, which can be parsed using the [getParameters] function.
+  ///
+  /// This function is passed the `Content-Type` header of the response as well
+  /// as its body as a UTF-8-decoded string. It should return a map in the same
+  /// format as the [standard JSON response][].
+  ///
+  /// [standard JSON response]: https://tools.ietf.org/html/rfc6749#section-5.1
+  DeviceAuthorizationGrant(
+    this.identifier,
+    this.deviceEndpoint,
+    this.tokenEndpoint, {
+    this.secret,
+    String? delimiter,
+    bool basicAuth = true,
+    http.Client? httpClient,
+    CredentialsRefreshedCallback? onCredentialsRefreshed,
+    Map<String, dynamic> Function(MediaType? contentType, String body)?
+        getParameters,
+  })  : _basicAuth = basicAuth,
+        _httpClient = httpClient ?? http.Client(),
+        _delimiter = delimiter ?? ' ',
+        _getParameters = getParameters ?? parseJsonParameters,
+        _onCredentialsRefreshed = onCredentialsRefreshed;
+
+  /// starts the Flow and returns [DeviceProperties]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<DeviceProperties> getDeviceCode(
+      {Iterable<String>? scopes, bool useSecret = false}) async {
+    if (_state != _State.initial) {
+      throw StateError('The device_code has already been generated.');
+    }
+
+    var scopeList = scopes?.toList() ?? <String>[];
+    _scopes = scopeList;
+
+    var headers = <String, String>{};
+    var body = <String, String>{};
+
+    var secret = this.secret;
+    if (_basicAuth && secret != null && useSecret) {
+      headers['Authorization'] = basicAuthHeader(identifier, secret);
+    } else {
+      // The ID is required for this request any time basic auth isn't being
+      // used, even if there's no actual client authentication to be done.
+      body['client_id'] = identifier;
+      if (secret != null && useSecret) body['client_secret'] = secret;
+    }
+
+    if (scopeList.isNotEmpty) body['scope'] = scopeList.join(_delimiter);
+
+    var startTime = DateTime.now();
+
+    var response =
+        await _httpClient.post(deviceEndpoint, headers: headers, body: body);
```

Comment:
Does the RFC specify any retry logic? If not is there a retry logic that we should consider implementing?

Won't most users of something like this want to have retry logic around the HTTP request?

Reply:
The RFC talks about polling. A standard interval of 5 seconds (according to RFC) is assumed, unless otherwise defined.

Reply:
I have implemented a basic logic for the query, but the user can also implement his own query logic